### ISSUE TITLE
on_*_callback unexpected behavior with `functools.partial`

### ISSUE THREAD

**Issue by patgarz** [apache/airflow#32158]
### Apache Airflow version

2.6.2

### What happened

Defining `on_*_callbacks` in Airflow 2.6.0+ using `functools.partial` (for example, to define a callable with args) causes unexpected behavior. Most notably, defining an `on_success_callback` that fails causes the process to crash(? I think), causing a failing heartbeat and marking the task as a zombie job: `ERROR - Detected zombie job`.

For `on_success_callback`, this most notably causes tasks to be marked as failure even though it was the callback that failed. However, this issue also persists in other erratic ways (likely due to the zombie job):

1. Logging presents a strange message, `'functools.partial' object has no attribute '__module__'`, rather than the actual callback exception
2. Configuring `on_failure_callback` as well as a partial-defined, failing `on_success_callback` causes the `on_failure_callback` to fire when the `on_success_callback` fails
3. Configuring `on_failure_callback` as a `partial` and a failing `email_on_failure` causes the failure callback to continuously trigger on a loop and the DAGRun never exits
4. HTTP logs cannot be retrieved

### What you think should happen instead

When using Airflow <=2.5.3, defining a callback with or without partial is identical. Likewise, defining a function in 2.6.0+ without `partial` causes the callback to process as expected:

- If the callback fails, the task status matches the task execution, not the callback status
- Callbacks do not initiate retries
- Exception text matches the callback exception

### How to reproduce

Sample DAG. Remove/comment/etc. various pieces to see how they interact (partials, callbacks, email_on_failure, etc).

```py
from datetime import timedelta
from functools import partial
import time
from airflow.decorators import dag, task
import pendulum
import logging


def my_callback(context):
    logging.info("success callback")
    raise Exception("exception text")


default_args = {
    "on_success_callback": partial(my_callback),
    "on_failure_callback": partial(my_callback),
    "email_on_failure": True,
    "email": "whatever@whatever.whatever",
}


@dag(
    dag_id="Start_and_Wait",
    schedule=None,
    start_date=pendulum.datetime(2022, 11, 15, tz="UTC"),
    catchup=False,
    default_args=default_args,
)
def wait_flow():
    @task
    def wait():
        time.sleep(1)

    @task
    def log():
        logging.info("done!")

    wait() >> log()


wait_flow()

```

### Operating System

Ubuntu 20.04

### Versions of Apache Airflow Providers

Experiencing issue in apache-airflow 2.6.0+
Not experiencing issue in apache-airflow <=2.5.3

Recommended / upstream constraints file in all cases

### Deployment

Other Docker-based deployment

### Deployment details

Presents in both LocalExecutor and CeleryExecutor

### Anything else

Expected execution in 2.5.3:
```txt
c336145d4ae9
*** Reading local file: /usr/local/airflow/shared/logs/dag_id=Start_and_Wait/run_id=manual__2023-06-26T20:07:31.314971+00:00/task_id=wait/attempt=1.log
[2023-06-26, 20:07:31 UTC] {taskinstance.py:1090} INFO - Dependencies all met for dep_context=non-requeueable deps ti=<TaskInstance: Start_and_Wait.wait manual__2023-06-26T20:07:31.314971+00:00 [queued]>
[2023-06-26, 20:07:31 UTC] {taskinstance.py:1090} INFO - Dependencies all met for dep_context=requeueable deps ti=<TaskInstance: Start_and_Wait.wait manual__2023-06-26T20:07:31.314971+00:00 [queued]>
[2023-06-26, 20:07:31 UTC] {taskinstance.py:1288} INFO - 
--------------------------------------------------------------------------------
[2023-06-26, 20:07:31 UTC] {taskinstance.py:1289} INFO - Starting attempt 1 of 1
[2023-06-26, 20:07:31 UTC] {taskinstance.py:1290} INFO - 
--------------------------------------------------------------------------------
[2023-06-26, 20:07:31 UTC] {taskinstance.py:1309} INFO - Executing <Task(_PythonDecoratedOperator): wait> on 2023-06-26 20:07:31.314971+00:00
[2023-06-26, 20:07:31 UTC] {standard_task_runner.py:55} INFO - Started process 901 to run task
[2023-06-26, 20:07:31 UTC] {standard_task_runner.py:82} INFO - Running: ['***', 'tasks', 'run', 'Start_and_Wait', 'wait', 'manual__2023-06-26T20:07:31.314971+00:00', '--job-id', '12', '--raw', '--subdir', 'DAGS_FOLDER/_local/wait.py', '--cfg-path', '/tmp/tmpmm0tra8e']
[2023-06-26, 20:07:31 UTC] {standard_task_runner.py:83} INFO - Job 12: Subtask wait
[2023-06-26, 20:07:32 UTC] {task_command.py:389} INFO - Running <TaskInstance: Start_and_Wait.wait manual__2023-06-26T20:07:31.314971+00:00 [running]> on host c336145d4ae9
[2023-06-26, 20:07:32 UTC] {taskinstance.py:1516} INFO - Exporting the following env vars:
AIRFLOW_CTX_DAG_EMAIL=***
AIRFLOW_CTX_DAG_OWNER=***
AIRFLOW_CTX_DAG_ID=Start_and_Wait
AIRFLOW_CTX_TASK_ID=wait
AIRFLOW_CTX_EXECUTION_DATE=2023-06-26T20:07:31.314971+00:00
AIRFLOW_CTX_TRY_NUMBER=1
AIRFLOW_CTX_DAG_RUN_ID=manual__2023-06-26T20:07:31.314971+00:00
[2023-06-26, 20:07:33 UTC] {python.py:177} INFO - Done. Returned value was: None
[2023-06-26, 20:07:33 UTC] {taskinstance.py:1327} INFO - Marking task as SUCCESS. dag_id=Start_and_Wait, task_id=wait, execution_date=20230626T200731, start_date=20230626T200731, end_date=20230626T200733
[2023-06-26, 20:07:33 UTC] {wait.py:9} INFO - success callback
[2023-06-26, 20:07:33 UTC] {taskinstance.py:1544} ERROR - Error when executing on_success callback
Traceback (most recent call last):
  File "/app/.local/lib/python3.9/site-packages/airflow/models/taskinstance.py", line 1542, in _run_finished_callback
    callback(context)
  File "/usr/local/airflow/dags/_local/wait.py", line 10, in my_callback
    raise Exception("exception text")
Exception: exception text
[2023-06-26, 20:07:33 UTC] {local_task_job.py:212} INFO - Task exited with return code 0
[2023-06-26, 20:07:33 UTC] {taskinstance.py:2596} INFO - 1 downstream tasks scheduled from follow-on schedule check
```

In 2.6.0+ (task marked as failed by scheduler? regardless of log here saying marked success):
```txt
5c688d484e65
*** Found local files:
***   * /usr/local/airflow/shared/logs/dag_id=Start_and_Wait/run_id=manual__2023-06-26T20:04:44.289200+00:00/task_id=wait/attempt=1.log
*** Could not read served logs: timed out
[2023-06-26, 20:04:45 UTC] {taskinstance.py:1103} INFO - Dependencies all met for dep_context=non-requeueable deps ti=<TaskInstance: Start_and_Wait.wait manual__2023-06-26T20:04:44.289200+00:00 [queued]>
[2023-06-26, 20:04:45 UTC] {taskinstance.py:1103} INFO - Dependencies all met for dep_context=requeueable deps ti=<TaskInstance: Start_and_Wait.wait manual__2023-06-26T20:04:44.289200+00:00 [queued]>
[2023-06-26, 20:04:45 UTC] {taskinstance.py:1308} INFO - Starting attempt 1 of 1
[2023-06-26, 20:04:45 UTC] {taskinstance.py:1327} INFO - Executing <Task(_PythonDecoratedOperator): wait> on 2023-06-26 20:04:44.289200+00:00
[2023-06-26, 20:04:45 UTC] {standard_task_runner.py:57} INFO - Started process 786 to run task
[2023-06-26, 20:04:45 UTC] {standard_task_runner.py:84} INFO - Running: ['***', 'tasks', 'run', 'Start_and_Wait', 'wait', 'manual__2023-06-26T20:04:44.289200+00:00', '--job-id', '14', '--raw', '--subdir', 'DAGS_FOLDER/_local/wait.py', '--cfg-path', '/tmp/tmpx798o7bj']
[2023-06-26, 20:04:45 UTC] {standard_task_runner.py:85} INFO - Job 14: Subtask wait
[2023-06-26, 20:04:45 UTC] {task_command.py:410} INFO - Running <TaskInstance: Start_and_Wait.wait manual__2023-06-26T20:04:44.289200+00:00 [running]> on host 5c688d484e65
[2023-06-26, 20:04:45 UTC] {taskinstance.py:1545} INFO - Exporting env vars: AIRFLOW_CTX_DAG_EMAIL='***' AIRFLOW_CTX_DAG_OWNER='***' AIRFLOW_CTX_DAG_ID='Start_and_Wait' AIRFLOW_CTX_TASK_ID='wait' AIRFLOW_CTX_EXECUTION_DATE='2023-06-26T20:04:44.289200+00:00' AIRFLOW_CTX_TRY_NUMBER='1' AIRFLOW_CTX_DAG_RUN_ID='manual__2023-06-26T20:04:44.289200+00:00'
[2023-06-26, 20:04:46 UTC] {python.py:183} INFO - Done. Returned value was: None
[2023-06-26, 20:04:46 UTC] {taskinstance.py:1345} INFO - Marking task as SUCCESS. dag_id=Start_and_Wait, task_id=wait, execution_date=20230626T200444, start_date=20230626T200445, end_date=20230626T200446
[2023-06-26, 20:04:46 UTC] {wait.py:9} INFO - success callback
[2023-06-26, 20:04:46 UTC] {standard_task_runner.py:104} ERROR - Failed to execute job 14 for task wait ('functools.partial' object has no attribute '__module__'; 786)
[2023-06-26, 20:04:46 UTC] {local_task_job_runner.py:225} INFO - Task exited with return code 1
[2023-06-26, 20:04:46 UTC] {taskinstance.py:2651} INFO - 0 downstream tasks scheduled from follow-on schedule check
```

Server logs:
```txt
***-local-webserver-local-1  | [2023-06-26T20:29:28.371+0000] {scheduler_job_runner.py:1688} WARNING - Failing (1) jobs without heartbeat after 2023-06-26 20:24:28.364681+00:00
***-local-webserver-local-1  | [2023-06-26T20:29:28.371+0000] {scheduler_job_runner.py:1698} ERROR - Detected zombie job: {'full_filepath': '/usr/local/airflow/dags/_local/wait.py', 'processor_subdir': '/usr/local/airflow/dags', 'msg': "{'DAG Id': 'Start_and_Wait', 'Task Id': 'wait', 'Run Id': 'manual__2023-06-26T20:29:06.100360+00:00', 'Hostname': 'ee1eef189157'}", 'simple_task_instance': <airflow.models.taskinstance.SimpleTaskInstance object at 0x7fa1e59eaac0>, 'is_failure_callback': True}
```

Failure to fetch logfile via HTTP:
```txt
***-local-webserver-local-1  | 172.23.0.1 - - [26/Jun/2023:20:15:14 +0000] "GET /api/v1/dags/Start_and_Wait/dagRuns/manual__2023-06-26T20:14:45.191590+00:00/taskInstances/wait/logs/1?full_content=false HTTP/1.1" 200 2618 "http://localhost:8080/dags/Start_and_Wait/grid?tab=logs&dag_run_id=manual__2023-06-26T20%3A14%3A45.191590%2B00%3A00&task_id=wait" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36"
***-local-webserver-local-1  | [2023-06-26T20:15:14.800+0000] {file_task_handler.py:522} ERROR - Could not read served logs
***-local-webserver-local-1  | Traceback (most recent call last):
***-local-webserver-local-1  |   File "/app/.local/lib/python3.9/site-packages/httpcore/_exceptions.py", line 10, in map_exceptions
***-local-webserver-local-1  |     yield
***-local-webserver-local-1  |   File "/app/.local/lib/python3.9/site-packages/httpcore/backends/sync.py", line 28, in read
***-local-webserver-local-1  |     return self._sock.recv(max_bytes)
***-local-webserver-local-1  | socket.timeout: timed out
```

What should happen (No `partial` usage):

```txt
281bdb0efae9
*** Found local files:
***   * /usr/local/airflow/shared/logs/dag_id=Start_and_Wait/run_id=manual__2023-06-26T21:29:36.991541+00:00/task_id=wait/attempt=1.log
[2023-06-26, 21:29:38 UTC] {taskinstance.py:1103} INFO - Dependencies all met for dep_context=non-requeueable deps ti=<TaskInstance: Start_and_Wait.wait manual__2023-06-26T21:29:36.991541+00:00 [queued]>
[2023-06-26, 21:29:38 UTC] {taskinstance.py:1103} INFO - Dependencies all met for dep_context=requeueable deps ti=<TaskInstance: Start_and_Wait.wait manual__2023-06-26T21:29:36.991541+00:00 [queued]>
[2023-06-26, 21:29:38 UTC] {taskinstance.py:1308} INFO - Starting attempt 1 of 2
[2023-06-26, 21:29:38 UTC] {taskinstance.py:1327} INFO - Executing <Task(_PythonDecoratedOperator): wait> on 2023-06-26 21:29:36.991541+00:00
[2023-06-26, 21:29:38 UTC] {standard_task_runner.py:57} INFO - Started process 511 to run task
[2023-06-26, 21:29:38 UTC] {standard_task_runner.py:84} INFO - Running: ['***', 'tasks', 'run', 'Start_and_Wait', 'wait', 'manual__2023-06-26T21:29:36.991541+00:00', '--job-id', '32', '--raw', '--subdir', 'DAGS_FOLDER/_local/wait.py', '--cfg-path', '/tmp/tmpsi6mlmbn']
[2023-06-26, 21:29:38 UTC] {standard_task_runner.py:85} INFO - Job 32: Subtask wait
[2023-06-26, 21:29:38 UTC] {task_command.py:410} INFO - Running <TaskInstance: Start_and_Wait.wait manual__2023-06-26T21:29:36.991541+00:00 [running]> on host 281bdb0efae9
[2023-06-26, 21:29:38 UTC] {taskinstance.py:1545} INFO - Exporting env vars: AIRFLOW_CTX_DAG_EMAIL='***' AIRFLOW_CTX_DAG_OWNER='***' AIRFLOW_CTX_DAG_ID='Start_and_Wait' AIRFLOW_CTX_TASK_ID='wait' AIRFLOW_CTX_EXECUTION_DATE='2023-06-26T21:29:36.991541+00:00' AIRFLOW_CTX_TRY_NUMBER='1' AIRFLOW_CTX_DAG_RUN_ID='manual__2023-06-26T21:29:36.991541+00:00'
[2023-06-26, 21:29:39 UTC] {python.py:183} INFO - Done. Returned value was: None
[2023-06-26, 21:29:39 UTC] {taskinstance.py:1345} INFO - Marking task as SUCCESS. dag_id=Start_and_Wait, task_id=wait, execution_date=20230626T212936, start_date=20230626T212938, end_date=20230626T212939
[2023-06-26, 21:29:39 UTC] {wait.py:10} INFO - success callback
[2023-06-26, 21:29:39 UTC] {taskinstance.py:1581} ERROR - Error when executing my_callback callback
Traceback (most recent call last):
  File "/app/.local/lib/python3.9/site-packages/airflow/models/taskinstance.py", line 1578, in _run_finished_callback
    callback(context)
  File "/usr/local/airflow/dags/_local/wait.py", line 11, in my_callback
    raise Exception("exception text")
Exception: exception text
[2023-06-26, 21:29:39 UTC] {local_task_job_runner.py:225} INFO - Task exited with return code 0
[2023-06-26, 21:29:39 UTC] {taskinstance.py:2651} INFO - 1 downstream tasks scheduled from follow-on schedule check
```

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by hussein-awala** [comment]
I just tested it on main branch, the tasks have the state `success`.

However, I have the callback exception in the task log, and from Airflow documentation:
> Callback functions are executed after tasks are completed. Errors in callback functions will show up in scheduler logs rather than task logs. By default, scheduler logs do not show up in the UI and instead can be found in $AIRFLOW_HOME/logs/scheduler/latest/PROJECT/DAG_FILE.py.log

So there is a bug which need to be fixed.

@potiuk do you have an idea which one we should fix, the documentation or the callback error handling?

**Comment by potiuk** [comment]
I think currently errors in the callback can show in various places - I think simply speaking, they will show up in the context they area called  - and it depends not only on the callback type but also sometimes on the context where they are executed (some callbacks are executed in DAG file processor for one.

I am not sure if we want to do anything to change it (might be difficult) - more like document it.

**Comment by ephraimbuddy** [comment]
I feel we should be able to give the same error message when using partial or not. Taking a look

**Comment by ephraimbuddy** [comment]
> I just tested it on main branch, the tasks have the state `success`.
> 
> However, I have the callback exception in the task log, and from Airflow documentation:
> 
> > Callback functions are executed after tasks are completed. Errors in callback functions will show up in scheduler logs rather than task logs. By default, scheduler logs do not show up in the UI and instead can be found in $AIRFLOW_HOME/logs/scheduler/latest/PROJECT/DAG_FILE.py.log
> 
> So there is a bug which need to be fixed.
> 
> @potiuk do you have an idea which one we should fix, the documentation or the callback error handling?

Hmm, same thing I got and I feel there's nothing to fix. When you have task level callbacks, you will see the execution log in both the task log and dag parsing logs. But when you have DAG level callback, you only see the logs in dag parsing.

@patgarz , I can't get the error below using your DAG:
```
ERROR - Failed to execute job 14 for task wait ('functools.partial' object has no attribute '__module__'; 786)
```

**Comment by patgarz** [comment]
Hey @ephraimbuddy , I concur with @hussein-awala --- it is fixed on `main`. I was only able to reproduce in 2.6.0, 2.6.1, and 2.6.2. Testing using `breeze` and the main I no longer encountered the issue. 

I assume that means it's fixed, but I suppose it could also represent a bug that's somehow added during packaging.

**Comment by patgarz** [comment]
Thanks; I did some additional testing and I am confident that the issue is resolved in >=2.7:

- Using `breeze` with `v2-6-test` branch produced results consistent with Airflow 2.6.x
- Using `breeze release-management prepare-airflow-package` and installing in a barebones image produced results consistent with `breeze` on `main`

I'll leave the issue open as I'm not sure whether there are still pending conversations re: the documentation. But I think the underlying issue is no longer present.

**Comment by potiuk** [comment]
Closing is best in this case.
